### ISSUE TITLE
Generic lookup type wrongly narrowed to `never` with strictNullChecks

### ISSUE THREAD

**Issue by mattmccutchen** [Microsoft/TypeScript#26130]
I ran across this while migrating my project to `strictNullChecks` and it's positively bizarre.

**TypeScript Version:**  master (50f442f)

**Search Terms:** mapped type narrow narrowing null strictNullChecks never

**Code** (with `strictNullChecks`)

```ts
let mappedObject: {[K in "foo"]: null | {x: string}} = {foo: {x: "hello"}};
declare function foo<T>(x: T): null | T;

function bar<K extends "foo">(key: K) {
  const element = foo(mappedObject[key]);
  if (element == null)
    return;
  const x = element.x;
}
```

**Expected behavior:** Compile without error.

**Actual behavior:** Error: Property 'x' does not exist on type 'never'.

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=let%20mappedObject%3A%20%7B%5BK%20in%20%22foo%22%5D%3A%20null%20%7C%20%7Bx%3A%20string%7D%7D%20%3D%20%7Bfoo%3A%20%7Bx%3A%20%22hello%22%7D%7D%3B%0D%0Adeclare%20function%20foo%3CT%3E(x%3A%20T)%3A%20null%20%7C%20T%3B%0D%0A%0D%0Afunction%20bar%3CK%20extends%20%22foo%22%3E(key%3A%20K)%20%7B%0D%0A%20%20const%20element%20%3D%20foo(mappedObject%5Bkey%5D)%3B%0D%0A%20%20if%20(element%20%3D%3D%20null)%0D%0A%20%20%20%20return%3B%0D%0A%20%20const%20x%20%3D%20element.x%3B%0D%0A%7D%0D%0A) (remember to enable strictNullChecks)

**Related Issues:** Didn't find any.

**Comment by RyanCavanaugh** [comment]
This is just weird. Writing
```ts
const x = element!.x;
```
eliminates the error, but there shouldn't be anything for `!` to take away

**Comment by RyanCavanaugh** [comment]
Not a regression FWIW

**Comment by weswigham** [comment]
> This is just weird. Writing
`const x = element!.x;`
eliminates the error, but there shouldn't be anything for ! to take away

FWIW `!`, as a cast, also resets a control flow-narrowed type to its declared type. We have an open issue to make it respect narrowing.

**Comment by mattmccutchen** [comment]
I found the problem: in `getAssignmentReducedType`, we have a declared type of `{x: string} | null` and an assigned type of `{foo: {x: string} | null}[K] | null`.  When we ask if the assigned type is maybe assignable to `{x: string}`, we decompose the first level of union and check if `{foo: {x: string} | null}[K]` is assignable to `{x: string}`, but it isn't because its constraint is `{x: string} | null`.  So we throw `{x: string}` out of the declared type and are left with `null`, and after the null check, we are left with `never`.

Using the comparable relation in place of `typeMaybeAssignableTo` fixes this issue.  I'm going to see what else breaks if I do that.

**Comment by mattmccutchen** [comment]
Proposed fix is in #26143.
